File: src/point.cpp
#include "point.hpp"

namespace vkmsm {
namespace {

Fp fp_double(const Fp& x) { return fp_add(x, x); }

// The BLS12-381 G1 curve constant b = 4, in Montgomery form.
const Fp& curve_b() {
    static const Fp b = [] {
        std::array<uint32_t, kLimbs> four{};
        four[0] = 4;
        return fp_from_plain_limbs(four);
    }();
    return b;
}

}  // namespace

PointJacobian point_infinity() {
    PointJacobian p;
    p.x = fp_one();
    p.y = fp_one();
    p.z = fp_zero();
    return p;
}

bool point_is_infinity(const PointJacobian& p) { return fp_is_zero(p.z); }

PointJacobian point_double(const PointJacobian& p) {
    if (point_is_infinity(p)) return p;

    // dbl-2009-l (EFD), specialized for a=0:
    //   A = X1^2, B = Y1^2, C = B^2
    //   D = 2*((X1+B)^2 - A - C)
    //   E = 3*A, F = E^2
    //   X3 = F - 2*D
    //   Y3 = E*(D - X3) - 8*C
    //   Z3 = 2*Y1*Z1
    Fp A = fp_mul(p.x, p.x);
    Fp B = fp_mul(p.y, p.y);
    Fp C = fp_mul(B, B);
    Fp X1plusB = fp_add(p.x, B);
    Fp D = fp_double(fp_sub(fp_sub(fp_mul(X1plusB, X1plusB), A), C));
    Fp E = fp_add(fp_double(A), A);
    Fp F = fp_mul(E, E);

    PointJacobian r;
    r.x = fp_sub(F, fp_double(D));
    Fp eightC = fp_double(fp_double(fp_double(C)));
    r.y = fp_sub(fp_mul(E, fp_sub(D, r.x)), eightC);
    r.z = fp_double(fp_mul(p.y, p.z));
    return r;
}

PointJacobian point_add(const PointJacobian& p1, const PointJacobian& p2) {
    if (point_is_infinity(p1)) return p2;
    if (point_is_infinity(p2)) return p1;

    Fp Z1Z1 = fp_mul(p1.z, p1.z);
    Fp Z2Z2 = fp_mul(p2.z, p2.z);
    Fp U1 = fp_mul(p1.x, Z2Z2);
    Fp U2 = fp_mul(p2.x, Z1Z1);
    Fp Z1cubed = fp_mul(Z1Z1, p1.z);
    Fp Z2cubed = fp_mul(Z2Z2, p2.z);
    Fp S1 = fp_mul(p1.y, Z2cubed);
    Fp S2 = fp_mul(p2.y, Z1cubed);

    if (fp_equal(U1, U2)) {
        // Same X: either the same point (-> double) or P == -Q (-> infinity).
        return fp_equal(S1, S2) ? point_double(p1) : point_infinity();
    }

    // add-2007-bl (EFD):
    //   H = U2-U1, I = (2H)^2, J = H*I
    //   r = 2*(S2-S1), V = U1*I
    //   X3 = r^2 - J - 2*V
    //   Y3 = r*(V-X3) - 2*S1*J
    //   Z3 = ((Z1+Z2)^2 - Z1Z1 - Z2Z2)*H
    Fp H = fp_sub(U2, U1);
    Fp twoH = fp_double(H);
    Fp I = fp_mul(twoH, twoH);
    Fp J = fp_mul(H, I);
    Fp rr = fp_double(fp_sub(S2, S1));
    Fp V = fp_mul(U1, I);

    PointJacobian result;
    result.x = fp_sub(fp_sub(fp_mul(rr, rr), J), fp_double(V));
    result.y = fp_sub(fp_mul(rr, fp_sub(V, result.x)), fp_double(fp_mul(S1, J)));
    Fp Z1plusZ2 = fp_add(p1.z, p2.z);
    result.z = fp_mul(fp_sub(fp_sub(fp_mul(Z1plusZ2, Z1plusZ2), Z1Z1), Z2Z2), H);
    return result;
}

bool point_on_curve(const PointJacobian& p) {
    if (point_is_infinity(p)) return true;
    Fp y2 = fp_mul(p.y, p.y);
    Fp x3 = fp_mul(fp_mul(p.x, p.x), p.x);
    Fp z2 = fp_mul(p.z, p.z);
    Fp z6 = fp_mul(fp_mul(z2, z2), z2);
    Fp rhs = fp_add(x3, fp_mul(curve_b(), z6));
    return fp_equal(y2, rhs);
}

bool point_equal(const PointJacobian& a, const PointJacobian& b) {
    bool a_inf = point_is_infinity(a);
    bool b_inf = point_is_infinity(b);
    if (a_inf || b_inf) return a_inf && b_inf;

    Fp Z1Z1 = fp_mul(a.z, a.z);
    Fp Z2Z2 = fp_mul(b.z, b.z);
    if (!fp_equal(fp_mul(a.x, Z2Z2), fp_mul(b.x, Z1Z1))) return false;

    Fp Z1cubed = fp_mul(Z1Z1, a.z);
    Fp Z2cubed = fp_mul(Z2Z2, b.z);
    return fp_equal(fp_mul(a.y, Z2cubed), fp_mul(b.y, Z1cubed));
}

}  // namespace vkmsm

File: jax_tpu/point.py
"""BLS12-381 G1 Jacobian point arithmetic in JAX, batched for TPU.

Mirrors ../src/point.cpp exactly: dbl-2009-l doubling, add-2007-bl
addition (both a=0 specializations), point == (X, Y, Z) with Z == 0
meaning infinity. Every function operates on arrays of shape
(..., LIMBS) per coordinate - the leading batch dimension is what
JAX/XLA parallelizes across.
"""

import jax.numpy as jnp

import fp


def is_infinity(z):
    """z: (..., LIMBS) uint32. Returns (...,) bool."""
    return jnp.all(z == 0, axis=-1)


def point_double(x, y, z):
    inf = is_infinity(z)

    A = fp.fp_mul(x, x)
    B = fp.fp_mul(y, y)
    C = fp.fp_mul(B, B)
    x1_plus_b = fp.fp_add(x, B)
    D = fp.fp_double(fp.fp_sub(fp.fp_sub(fp.fp_mul(x1_plus_b, x1_plus_b), A), C))
    E = fp.fp_add(fp.fp_double(A), A)
    F = fp.fp_mul(E, E)

    x3 = fp.fp_sub(F, fp.fp_double(D))
    eight_c = fp.fp_double(fp.fp_double(fp.fp_double(C)))
    y3 = fp.fp_sub(fp.fp_mul(E, fp.fp_sub(D, x3)), eight_c)
    z3 = fp.fp_double(fp.fp_mul(y, z))

    # Infinity stays infinity.
    x3 = jnp.where(inf[..., None], x, x3)
    y3 = jnp.where(inf[..., None], y, y3)
    z3 = jnp.where(inf[..., None], z, z3)
    return x3, y3, z3


def point_add(x1, y1, z1, x2, y2, z2):
    inf1 = is_infinity(z1)
    inf2 = is_infinity(z2)

    z1z1 = fp.fp_mul(z1, z1)
    z2z2 = fp.fp_mul(z2, z2)
    u1 = fp.fp_mul(x1, z2z2)
    u2 = fp.fp_mul(x2, z1z1)
    z1_cubed = fp.fp_mul(z1z1, z1)
    z2_cubed = fp.fp_mul(z2z2, z2)
    s1 = fp.fp_mul(y1, z2_cubed)
    s2 = fp.fp_mul(y2, z1_cubed)

    same_x = jnp.all(u1 == u2, axis=-1)
    same_y = jnp.all(s1 == s2, axis=-1)

    # Generic addition (add-2007-bl); only meaningful where !same_x, but
    # computed unconditionally (branch-free, the natural style for
    # vectorized/TPU code) and then selected via jnp.where below.
    h = fp.fp_sub(u2, u1)
    two_h = fp.fp_double(h)
    i_ = fp.fp_mul(two_h, two_h)
    j_ = fp.fp_mul(h, i_)
    r = fp.fp_double(fp.fp_sub(s2, s1))
    v = fp.fp_mul(u1, i_)

    x3_add = fp.fp_sub(fp.fp_sub(fp.fp_mul(r, r), j_), fp.fp_double(v))
    y3_add = fp.fp_sub(fp.fp_mul(r, fp.fp_sub(v, x3_add)), fp.fp_double(fp.fp_mul(s1, j_)))
    z1_plus_z2 = fp.fp_add(z1, z2)
    z3_add = fp.fp_mul(fp.fp_sub(fp.fp_sub(fp.fp_mul(z1_plus_z2, z1_plus_z2), z1z1), z2z2), h)

    # Doubling result, for the same_x & same_y (P == Q) case.
    x3_dbl, y3_dbl, z3_dbl = point_double(x1, y1, z1)

    zero = fp.fp_zero(x1.shape[:-1])
    # P == -Q (same_x, !same_y) -> infinity.
    x3 = jnp.where(same_x[..., None], jnp.where(same_y[..., None], x3_dbl, zero), x3_add)
    y3 = jnp.where(same_x[..., None], jnp.where(same_y[..., None], y3_dbl, zero), y3_add)
    z3 = jnp.where(same_x[..., None], jnp.where(same_y[..., None], z3_dbl, zero), z3_add)

    # inf1 -> return P2; inf2 -> return P1 (checked last so they win).
    x3 = jnp.where(inf2[..., None], x1, jnp.where(inf1[..., None], x2, x3))
    y3 = jnp.where(inf2[..., None], y1, jnp.where(inf1[..., None], y2, y3))
    z3 = jnp.where(inf2[..., None], z1, jnp.where(inf1[..., None], z2, z3))
    return x3, y3, z3


def points_equal(x1, y1, z1, x2, y2, z2):
    """Affine-equivalence check (X1*Z2^2 == X2*Z1^2 and Y1*Z2^3 == Y2*Z1^3),
    so differently-Z-scaled Jacobian reps of the same point compare equal."""
    inf1 = is_infinity(z1)
    inf2 = is_infinity(z2)
    both_inf = inf1 & inf2

    z1z1 = fp.fp_mul(z1, z1)
    z2z2 = fp.fp_mul(z2, z2)
    lhs_x = fp.fp_mul(x1, z2z2)
    rhs_x = fp.fp_mul(x2, z1z1)
    x_eq = jnp.all(lhs_x == rhs_x, axis=-1)

    lhs_y = fp.fp_mul(y1, fp.fp_mul(z2z2, z2))
    rhs_y = fp.fp_mul(y2, fp.fp_mul(z1z1, z1))
    y_eq = jnp.all(lhs_y == rhs_y, axis=-1)

    neither_inf = ~inf1 & ~inf2
    return jnp.where(inf1 | inf2, both_inf, x_eq & y_eq & neither_inf)
